src/rust/src/backend/kdf.rs

```diff
@@ -1696,11 +1696,321 @@ impl ConcatKdfHmac {
+        ));
+    }
+
+    Ok(KbkdfParams {
+        rlen,
+        llen,
+        location: rust_location,
+        label,
+        context,
+        fixed,
+        break_location,
+    })
+}
+
+impl KbkdfHmac {
+    fn derive_into_buffer(
+        &mut self,
+        py: pyo3::Python<'_>,
+        key_material: &[u8],
+        output: &mut [u8],
+    ) -> CryptographyResult<usize> {
+        if self.used {
+            return Err(exceptions::already_finalized_error());
+        }
+        self.used = true;
+
+        if output.len() != self.length {
+            return Err(CryptographyError::from(
+                pyo3::exceptions::PyValueError::new_err(format!(
+                    "buffer must be {} bytes",
+                    self.length
+                )),
+            ));
+        }
+
+        let algorithm_bound = self.algorithm.bind(py);
+        let digest_size = algorithm_bound
+            .getattr(pyo3::intern!(py, "digest_size"))?
+            .extract::<usize>()?;
+
+        let fixed = self.generate_fixed_input(py)?;
+
+        let (data_before_ctr, data_after_ctr) = match &self.params.location {
+            CounterLocation::BeforeFixed => (&b""[..], &fixed[..]),
+            CounterLocation::AfterFixed => (&fixed[..], &b""[..]),
+            CounterLocation::MiddleFixed => {
+                // We validate break_location is Some when counter_location is MiddleFixed
+                // in the validate function
+                let break_loc = self.params.break_location.unwrap();
+                if break_loc > fixed.len() {
+                    return Err(CryptographyError::from(
+                        pyo3::exceptions::PyValueError::new_err(
+                            "break_location offset > len(fixed)",
+                        ),
+                    ));
+                }
+                (&fixed[..break_loc], &fixed[break_loc..])
+            }
+        };
+
+        let mut pos = 0usize;
+        let rounds = self.length.div_ceil(digest_size);
+        for i in 1..=rounds {
+            let mut hmac = Hmac::new_bytes(py, key_material, algorithm_bound)?;
+
+            let counter = int_to_bytes(i, self.params.rlen);
+            hmac.update_bytes(data_before_ctr)?;
+            hmac.update_bytes(&counter)?;
+            hmac.update_bytes(data_after_ctr)?;
+
+            let result = hmac.finalize_bytes()?;
+
+            let copy_len = (self.length - pos).min(digest_size);
+            output[pos..pos + copy_len].copy_from_slice(&result[..copy_len]);
+            pos += copy_len;
+        }
+
+        Ok(self.length)
+    }
+
+    fn generate_fixed_input(&self, py: pyo3::Python<'_>) -> CryptographyResult<Vec<u8>> {
+        if let Some(ref fixed_data) = self.params.fixed {
+            return Ok(fixed_data.as_bytes(py).to_vec());
+        }
+
+        // llen will exist if fixed data is not provided
+        let l_val = int_to_bytes(self.length * 8, self.params.llen.unwrap());
```

Comment:
We don't ever verify the range of llen, but `int_to_bytes` will panic if it's too big: https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=a4c1b2f7067040c7a9f906a14fe15f0b